### ISSUE TITLE
BrowserRenderer event listeners break with Onsen UI

### ISSUE THREAD

**Issue by taskbit** [aspnet/Blazor#422]
Using Onsen UI breaks the Blazor onclick listener.

**Index.html**
```html
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8" />
  <title>BlazorOnsenTest</title>
  <base href="/" />
  <link rel="stylesheet" href="https://unpkg.com/onsenui/css/onsenui.css">
  <link rel="stylesheet" href="https://unpkg.com/onsenui/css/onsen-css-components.min.css">
</head>
<body>
  <app>Loading...</app>
  <script src="https://unpkg.com/onsenui/js/onsenui.min.js"></script>
  <script type="blazor-boot">
  </script>
</body>
</html>

```

**Example1 (MainLayout.cshtml)**
```html
@using Microsoft.AspNetCore.Blazor
@implements ILayoutComponent

<ons-card>
  <div class="title">
    Title
  </div>
  <div class="content">
    <ons-list>
      <ons-list-header>Caption</ons-list-header>
      <ons-list-item>Item1</ons-list-item>
      <ons-list-item>  <button @onclick(Clicked)>Item2</button>  </ons-list-item>
      <ons-list-item>Item3</ons-list-item>
    </ons-list>
  </div>
</ons-card>

@functions {
public RenderFragment Body { get; set; }
public void Clicked(){}
}
```

Clicking the button results in:
```javascript
MonoPlatform.ts:85 Uncaught Error: Microsoft.AspNetCore.Blazor.Browser.Interop.JavaScriptException: Cannot read property 'removeEventListener' of undefined
TypeError: Cannot read property 'removeEventListener' of undefined
    at BrowserRenderer.applyAttribute (BrowserRenderer.ts:182)
    at BrowserRenderer.applyEdits (BrowserRenderer.ts:57)
    at BrowserRenderer.updateComponent (BrowserRenderer.ts:25)
    at renderBatch (Renderer.ts:45)
    at Array.<anonymous> (mono.js:1)
    at _emscripten_asm_const_iiiiii (mono.js:1)
    at :45654/wasm-function[2609]:25
    at :45654/wasm-function[3361]:44
    at :45654/wasm-function[3368]:408
    at :45654/wasm-function[3369]:223
  at Microsoft.AspNetCore.Blazor.Browser.Interop.RegisteredFunction.InvokeUnmarshalled[T0,T1,T2,TRes] (:45654/System.String identifier, T0 arg0, T1 arg1, T2 arg2) <0x16a36a8 + 0x00054> in <aa678c00565443a1a03daf9fa7da0de1>:0 
  at Microsoft.AspNetCore.Blazor.Browser.Interop.RegisteredFunction.InvokeUnmarshalled[T0,T1,TRes] (:45654/System.String identifier, T0 arg0, T1 arg1) <0x16a35c0 + 0x00022> in <aa678c00565443a1a03daf9fa7da0de1>:0 
  at Microsoft.AspNetCore.Blazor.Browser.Rendering.BrowserRenderer.UpdateDisplay (:45654/Microsoft.AspNetCore.Blazor.Rendering.RenderBatch batch) <0x16a34d8 + 0x00020> in <aa678c00565443a1a03daf9fa7da0de1>:0 
  at :45654/Microsoft.AspNetCore.Blazor.Rendering.Renderer.ProcessRenderQueue () <0x15e02b0 + 0x00094> in <6f55c2fa30114054ab77a67cb3b359ec>:0 
  at Microsoft.AspNetCore.Blazor.Rendering.Renderer.DispatchEvent (:45654/System.Int32 componentId, System.Int32 eventHandlerId, Microsoft.AspNetCore.Blazor.UIEventArgs eventArgs) <0x1707f10 + 0x00068> in <6f55c2fa30114054ab77a67cb3b359ec>:0 
  at Microsoft.AspNetCore.Blazor.Browser.Rendering.BrowserRenderer.DispatchBrowserEvent (:45654/System.Int32 componentId, System.Int32 eventHandlerId, Microsoft.AspNetCore.Blazor.UIEventArgs eventArgs) <0x1707da8 + 0x00020> in <aa678c00565443a1a03daf9fa7da0de1>:0 
  at Microsoft.AspNetCore.Blazor.Browser.Rendering.BrowserRendererEventDispatcher.DispatchEvent (:45654/System.String eventDescriptorJson, System.String eventArgsJson) <0x16a47f8 + 0x0005a> in <aa678c00565443a1a03daf9fa7da0de1>:0 
    at BrowserRenderer.applyAttribute (BrowserRenderer.ts:182)
    at BrowserRenderer.applyEdits (BrowserRenderer.ts:57)
    at BrowserRenderer.updateComponent (BrowserRenderer.ts:25)
    at renderBatch (Renderer.ts:45)
    at Array.<anonymous> (mono.js:1)
    at _emscripten_asm_const_iiiiii (mono.js:1)
    at :45654/wasm-function[2609]:25
    at :45654/wasm-function[3361]:44
    at :45654/wasm-function[3368]:408
    at :45654/wasm-function[3369]:223
  at Microsoft.AspNetCore.Blazor.Browser.Interop.RegisteredFunction.InvokeUnmarshalled[T0,T1,T2,TRes] (:45654/System.String identifier, T0 arg0, T1 arg1, T2 arg2) <0x16a36a8 + 0x00054> in <aa678c00565443a1a03daf9fa7da0de1>:0 
  at Microsoft.AspNetCore.Blazor.Browser.Interop.RegisteredFunction.InvokeUnmarshalled[T0,T1,TRes] (:45654/System.String identifier, T0 arg0, T1 arg1) <0x16a35c0 + 0x00022> in <aa678c00565443a1a03daf9fa7da0de1>:0 
  at Microsoft.AspNetCore.Blazor.Browser.Rendering.BrowserRenderer.UpdateDisplay (:45654/Microsoft.AspNetCore.Blazor.Rendering.RenderBatch batch) <0x16a34d8 + 0x00020> in <aa678c00565443a1a03daf9fa7da0de1>:0 
  at :45654/Microsoft.AspNetCore.Blazor.Rendering.Renderer.ProcessRenderQueue () <0x15e02b0 + 0x00094> in <6f55c2fa30114054ab77a67cb3b359ec>:0 
  at Microsoft.AspNetCore.Blazor.Rendering.Renderer.DispatchEvent (:45654/System.Int32 componentId, System.Int32 eventHandlerId, Microsoft.AspNetCore.Blazor.UIEventArgs eventArgs) <0x1707f10 + 0x00068> in <6f55c2fa30114054ab77a67cb3b359ec>:0 
  at Microsoft.AspNetCore.Blazor.Browser.Rendering.BrowserRenderer.DispatchBrowserEvent (:45654/System.Int32 componentId, System.Int32 eventHandlerId, Microsoft.AspNetCore.Blazor.UIEventArgs eventArgs) <0x1707da8 + 0x00020> in <aa678c00565443a1a03daf9fa7da0de1>:0 
  at Microsoft.AspNetCore.Blazor.Browser.Rendering.BrowserRendererEventDispatcher.DispatchEvent (:45654/System.String eventDescriptorJson, System.String eventArgsJson) <0x16a47f8 + 0x0005a> in <aa678c00565443a1a03daf9fa7da0de1>:0 
    at Object.callMethod (MonoPlatform.ts:85)
    at raiseEvent (BrowserRenderer.ts:310)
    at HTMLButtonElement.listener [as _blazorClickListener] (BrowserRenderer.ts:183)
```

**Example2 (MainLayout.cshtml):**
```html
@using Microsoft.AspNetCore.Blazor
@implements ILayoutComponent

<ons-page id="helloworld-page">
  <p style="text-align: center">
    <ons-button @onclick(Clicked)>Click me!</ons-button>
  </p>
</ons-page>

@functions {
public RenderFragment Body { get; set; }
public void Clicked() { }
}
```
Clicking the button twice results in:
```javascript
MonoPlatform.ts:85 Uncaught Error: System.ArgumentException: There is no event handler with ID 1
  at Microsoft.AspNetCore.Blazor.Rendering.Renderer.DispatchEvent (:45654/System.Int32 componentId, System.Int32 eventHandlerId, Microsoft.AspNetCore.Blazor.UIEventArgs eventArgs) <0x1708008 + 0x00088> in <6f55c2fa30114054ab77a67cb3b359ec>:0 
  at Microsoft.AspNetCore.Blazor.Browser.Rendering.BrowserRenderer.DispatchBrowserEvent (:45654/System.Int32 componentId, System.Int32 eventHandlerId, Microsoft.AspNetCore.Blazor.UIEventArgs eventArgs) <0x1705978 + 0x00020> in <aa678c00565443a1a03daf9fa7da0de1>:0 
  at Microsoft.AspNetCore.Blazor.Browser.Rendering.BrowserRendererEventDispatcher.DispatchEvent (:45654/System.String eventDescriptorJson, System.String eventArgsJson) <0x16a4128 + 0x0005a> in <aa678c00565443a1a03daf9fa7da0de1>:0 
  at Microsoft.AspNetCore.Blazor.Rendering.Renderer.DispatchEvent (:45654/System.Int32 componentId, System.Int32 eventHandlerId, Microsoft.AspNetCore.Blazor.UIEventArgs eventArgs) <0x1708008 + 0x00088> in <6f55c2fa30114054ab77a67cb3b359ec>:0 
  at Microsoft.AspNetCore.Blazor.Browser.Rendering.BrowserRenderer.DispatchBrowserEvent (:45654/System.Int32 componentId, System.Int32 eventHandlerId, Microsoft.AspNetCore.Blazor.UIEventArgs eventArgs) <0x1705978 + 0x00020> in <aa678c00565443a1a03daf9fa7da0de1>:0 
  at Microsoft.AspNetCore.Blazor.Browser.Rendering.BrowserRendererEventDispatcher.DispatchEvent (:45654/System.String eventDescriptorJson, System.String eventArgsJson) <0x16a4128 + 0x0005a> in <aa678c00565443a1a03daf9fa7da0de1>:0 
    at Object.callMethod (MonoPlatform.ts:85)
    at raiseEvent (BrowserRenderer.ts:310)
    at HTMLElement.listener [as _blazorClickListener] (BrowserRenderer.ts:183)
callMethod @ MonoPlatform.ts:85
raiseEvent @ BrowserRenderer.ts:310
listener @ BrowserRenderer.ts:183
```

**Comment by SteveSandersonMS** [comment]
Thanks for letting us know!

I strongly suspect this is because we haven't yet implemented #433, but will check this once we have done so.

**Comment by SteveSandersonMS** [comment]
OK, I've investigated, and the reason this doesn't work is that OnsenUI is mutating the DOM directly in such a way that Blazor has no way to know how the older DOM structure relates to the new one. So after OnsenUI has mutated the DOM, subsequent diffs can no longer be applied because the DOM elements have been moved around.

The same would make OnsenUI impossible to use with React and Vue (and possibly Angular 2+, didn't check), except that they've published a special set of React and Vue-specific components for OnsenUI (https://onsen.io/v2/guide/react/#react and https://onsen.io/v2/guide/vue/#download-vue-bindings).

So if Onsen UI wanted to work with Blazor, it could do so, but they would have to publish Blazor-specific components just like they have done for React and Vue.

Basically, their approach of mutating the DOM directly will always need special-case support for frameworks that rely on diffing.

**Comment by JCKodel** [comment]
@SteveSandersonMS: I was wondering if Blazor could be used with Onsen UI and got here.
I would hapilly create a Onsen UI Blazor adapter, but I'm not sure how.
This is the source code of Onsen UI Vue adapter for a button (same component @taskbit is using):

```
<template>
  <ons-button v-on="unrecognizedListeners" :modifier="normalizedModifier">
    <slot></slot>
  </ons-button>
</template>

<script>
  /* This file was generated automatically by 'generate-components' task in bindings/vue/gulpfile.babel.js */
  import 'onsenui/esm/elements/ons-button';
  import { deriveEvents, modifier } from '../mixins';
  export default {
    name: 'v-ons-button',
    mixins: [deriveEvents, modifier]
  };
</script>
```

deriveEvents:
```
const deriveEvents = {
  computed: {
    unrecognizedListeners() {
      const name = camelize('-' + this.$options._componentTag.slice(6));
      return Object.keys(this.$listeners || {})
        .filter(k => (this.$ons.elements[name].events || []).indexOf(k) === -1)
        .reduce((r, k) => {
          r[k] = this.$listeners[k];
          return r;
        }, {});
    }
  },

  mounted() {
    this._handlers = {};

    (this.$el.constructor.events || []).forEach(key => {
      this._handlers[eventToHandler(key)] = event => {
        // Filter events from different components with the same name
        if (event.target === this.$el || !/^ons-/i.test(event.target.tagName)) {
          this.$emit(key, event);
        }
      };
      this.$el.addEventListener(key, this._handlers[eventToHandler(key)]);
    });
  },

  beforeDestroy() {
    Object.keys(this._handlers).forEach(key => {
      this.$el.removeEventListener(key, this._handlers[key]);
    });
    this._handlers = null;
  }
};
```

This binding only binds the Onsen UI evento to the Vue's this.$emit event emitter. But Blazor does not have a event emitter, right? It suppose to work with native <button onclick/> event (I think there's something here I'm not seeing, and is the key to solve this).
